Single Experiment Content Type

laceysanderson opened the pull request:
**Issue #25**

## Motivation
We have determined that our original approach of having multiple Experiment Type content types (i.e. Field Experiment, Greenhouse Experiment, Biochemical Assay) does not work well since many experiments may include components of two experiment types. For example, the microbiome work involves sampling from a field experiment and then sequencing to determine the species composition. Another example of this would be a food quality experiment where we need to describe the biochem assay and the way the seeds were obtained.

## What does this PR do?

As such, this PR brings together all the fields from the original Experiment Type content types into a single content type. It also strives to organize these fields in a less overwhelming way and only have one copy of any common fields (i.e. name, objectives, summary).

More specifically, this PR

1. Removes the individual experiment type content types from `tripal.tripalentitytype_collection.trpcultivate_experiments.yml` in favour of a single `research_experiment` content type.
2. Rearranges, names and combines any common fields across the 4 content types into a single content type in `tripal.tripalfield_collection.trpcultivate_experiments.yml` making sure not to loose any experiment type specific instructions.
3. Adds a default layout for the display (page view) of an experiment in `tripal_layout.tripal_layout_default_view.trpcultexperiment_layout.yml` to automate configuring the order of the many fields.
4. Adds a default layout for the form (edit) of an experiment in `tripal_layout.tripal_layout_default_form.trpcultexperiment_layout.yml` to automate configuring the order of the many fields.
5. Works to provide a consistent and intuitive organization of fields for both the experiment form and display.
6. Changes the "Germplasm Species" field to a "Germplasm Scientific Name" field to handle the situation where there are multiple genus'

## Testing

### Manual Testing

1. Create a docker on this branch.
2. Go to the content type listing and confirm we only have a single research experiment content type.
  <img width="1385" src="https://github.com/user-attachments/assets/95dfddfd-ebd4-43c9-805a-97c9a9b0d3bc">
3. Also go to the add content listing and confirm the same plus check that the definition for the research experiment vs. study is helpful, clear, accurate.
  <img width="1401" src="https://github.com/user-attachments/assets/95e15955-84e1-45e1-8cab-22a8ca599d5e">
4. Manage the form display of the Research Experiment content type, click the "Apply Tripal Layout" button and verify all the fields are present and organized in an intuitive manner.
  <img src="https://github.com/user-attachments/assets/5e3cfe43-2fc3-4f51-a7b0-ab5621d23730">
5. Manage the page display of the Research Experiment content type, click the "Apply Tripal Layout" button and verify all the fields are present and organized in an intuitive manner.
  <img src="https://github.com/user-attachments/assets/6fe3a775-3acc-493b-a985-5605e1e40dfc">
7. Create an experiment by filling out a variety of different fields and confirm it saves without error, the form is intuitive and functional, the help text is accurate and helpful, etc.
  <img src="https://github.com/user-attachments/assets/bdd88c4f-2dfa-4f3b-8bd2-d56ebe3e652b">

### Automated Testing

The automated tests are updated and just require #1993 before they will pass.

laceysanderson committed:
First attempt at combining the experiment types.

laceysanderson committed:
Fix a few more experiment fields.

laceysanderson committed:
Add a default layout for viewing an experiment.

laceysanderson committed:
Map out a form layout -not functional; dependant on update in core.

laceysanderson committed:
Switch to new format.

laceysanderson committed:
Move name help text to fieldset.

laceysanderson committed:
Small fixes to formatting of name guidelines.

laceysanderson committed:
Don't hide title/uid.

laceysanderson committed:
Improve experimental design section.

laceysanderson committed:
Make content types test a kernel test and fix counts with new changes.

laceysanderson committed:
More finetuning of the form layout.

laceysanderson committed:
Capture Tripal Layout dependancy.

carolyncaron commented on config/install/tripal.tripalentitytype_collection.trpcultivate_experiments.yml:
```diff
@@ -27,34 +27,10 @@ content_types:
-        id: biochem_experiment
+    -   label: Research Experiment
+        term: 'SIO:000994'
+        help_text: Use to describe a single experimental design with the goal of collecting data and answering specific hypothesis. Usually multiple experiments will be grouped together under a research study in order to make more broad conclusions.
```
```suggestion
        help_text: Use to describe a single experimental design with the goal of collecting data and answering a specific hypothesis. Usually multiple experiments will be grouped together under a research study in order to make more broad conclusions.
```

carolyncaron commented on config/install/tripal_layout.tripal_layout_default_form.trpcultexperiment_layout.yml:
```diff
@@ -0,0 +1,153 @@
+      - id: experiment_people
+        type: details
+        label: "Researchers"
+        description: "The fields are used to ensure everyone involved in the experiment recieves credit for the work they did."
```
```suggestion
        description: "The fields are used to ensure everyone involved in the experiment receives credit for the work they did."
```

carolyncaron commented on config/install/tripal_layout.tripal_layout_default_form.trpcultexperiment_layout.yml:
```diff
@@ -0,0 +1,153 @@
+      - id: experiment_funding
+        type: details
+        label: "Funding"
+        description: "The fields ensure that all funding organizations behind this research recieve credit for funding this work. You may need to talk your PI for these details."
```
```suggestion
        description: "The fields ensure that all funding organizations behind this research receive credit for funding this work. You may need to talk your PI for these details."
```

carolyncaron commented on config/install/tripal_layout.tripal_layout_default_form.trpcultexperiment_layout.yml:
```diff
@@ -0,0 +1,153 @@
+      - id: experiment_attribution
+        type: details
+        label: "Attribution"
+        description: "The fields are used to ensure everyone involved in the experiment recieves credit for the work they did."
```
This might be more obvious once I can test in the UI, but why does this share the same description as line 114?

```suggestion
        description: "The fields are used to ensure everyone involved in the experiment receives credit for the work they did."
```

carolyncaron reviewed (commented):
Just a quick pass at a code review with some suggestions to fix typos. Manual review still to come!

laceysanderson committed:
Apply suggestions from code review

Co-authored-by: Carolyn T Caron <carolyn.caron@usask.ca>

laceysanderson commented on config/install/tripal_layout.tripal_layout_default_form.trpcultexperiment_layout.yml:
```diff
@@ -0,0 +1,153 @@
+      - id: experiment_attribution
+        type: details
+        label: "Attribution"
+        description: "The fields are used to ensure everyone involved in the experiment receives credit for the work they did."
```
```suggestion
        description: "These fields are focused on giving credit to all researchers, organizations and funding related to this experiment. Additionally it provides information to people using the data from this experiment on the appropriate citation to use and any restrictions that apply."
```

laceysanderson committed:
Be more general for attribution.

carolyncaron commented on config/install/tripal.tripalfield_collection.trpcultivate_experiments.yml:
```diff
@@ -1758,24 +614,28 @@ fields:
+
+    -   name: exp_gchamber_locations
+        content_type: research_experiment
+        label: Grownth Chamber Locations
```
```suggestion
        label: Growth Chamber Locations
```

carolyncaron commented on config/install/tripal.tripalfield_collection.trpcultivate_experiments.yml:
```diff
@@ -450,1306 +476,136 @@ fields:
+        content_type: research_experiment
+        label: Field Experimental Design
+        type: chado_property_type_default
+        description: Describe the design of your experiment in detail. More specifically, include information on replication, randomization and blocking. Also include the location-year combinations you intend to collect data in. Make sure these match those specified in "Field Locations" and "Experiment Years".
```
```suggestion
        description: Describe the design of your experiment in detail. More specifically, include information on replication, randomization and blocking. Also include the location-year combinations you intend to collect data in, which should be consistent with "Field Locations" and "Experiment Years".
```

carolyncaron commented on config/install/tripal.tripalfield_collection.trpcultivate_experiments.yml:
```diff
@@ -450,1306 +476,136 @@ fields:
+        label: Experimental Design
         type: chado_property_type_default
-        description: The genus of the germplasm that data was collected for (i.e. that the experiment was done on). If data was collected for multiple genuses then each genus should be its own entry.
+        description: Describe the design of your experiment in detail. More specifically, include information on replication, randomization and blocking. Also include the location-year combinations you intend to collect data in. Make sure these match those specifieid in "Greenhouse Locations" and "Experiment Years".
```
```suggestion
        description: Describe the design of your experiment in detail. More specifically, include information on replication, randomization and blocking. Also include the location-year combinations you intend to collect data in, which should be consistent with "Greenhouse Locations" and "Experiment Years".
```

carolyncaron commented on config/install/tripal.tripalfield_collection.trpcultivate_experiments.yml:
```diff
@@ -1810,26 +670,24 @@ fields:
+        label: Experimental Design
         type: chado_property_type_default
-        description: 'List the full names of the core measurements being taken in this experiment. For example, if you are exploring protein concentration in lentils, then a core measurement could be: "Alanine content (% on dry basis) in whole lentils".'
+        description: Describe the design of your experiment in detail. More specifically, include information on replication, randomization and blocking. Also include the location-year combinations you intend to collect data in. Make sure these match those specified in "Growth Chamber Locations" and "Experiment Years".
```
```suggestion
        description: Describe the design of your experiment in detail. More specifically, include information on replication, randomization and blocking. Also include the location-year combinations you intend to collect data in, which should be consistent with "Growth Chamber Locations" and "Experiment Years".
```

carolyncaron commented on config/install/tripal_layout.tripal_layout_default_form.trpcultexperiment_layout.yml:
```diff
@@ -0,0 +1,153 @@
+      - id: experiment_people
+        type: details
+        label: "Researchers"
+        description: "The fields are used to ensure everyone involved in the experiment receives credit for the work they did."
```
```suggestion
        description: "These fields are used to ensure everyone involved in the experiment receives credit for the work they did."
```

carolyncaron commented on config/install/tripal_layout.tripal_layout_default_form.trpcultexperiment_layout.yml:
```diff
@@ -0,0 +1,153 @@
+      - id: experiment_funding
+        type: details
+        label: "Funding"
+        description: "The fields ensure that all funding organizations behind this research receive credit for funding this work. You may need to talk your PI for these details."
```
```suggestion
        description: "These fields ensure that all funding organizations behind this research receive credit for funding this work. You may need to talk your PI for these details."
```

carolyncaron reviewed (changes requested):
The following are thoughts I had while filling out the Research Experiment form, but are not specific to the updates made so I can't make suggestions in the code.

- I think it would be helpful to bold text that reminds the user to add another item when there are multiple:
   - Experiment Years:
   ![Screenshot 2024-10-09 at 11 53 02 AM](https://github.com/user-attachments/assets/8468b931-955c-4c5a-880c-0b26dffcae5d)
   - Germplasm Genus:
   ![Screenshot 2024-10-09 at 11 54 23 AM](https://github.com/user-attachments/assets/b05a0efd-a22e-495f-8f36-dd9510dbbba8)
   - Germplasm Species:
   ![Screenshot 2024-10-09 at 11 56 00 AM](https://github.com/user-attachments/assets/ef9ec75f-f843-42b8-bed8-7d4290d87bbd)
   - Data License(s):
   ![Screenshot 2024-10-09 at 12 00 59 PM](https://github.com/user-attachments/assets/6074f255-d2d4-447e-a673-2ad7ccc76ef9)

- Regarding Germplasm genus and species - what happens when we have more than one genus (`Lens` + `Phaseolus`) and then multiple species for each?

carolyncaron reviewed (commented):
I was also unable to save my experiment 😢 I filled out as many fields as I could for a Field Experiment. This also occurred if I only filled the `Name` field.

![Screenshot 2024-10-09 at 12 46 16 PM](https://github.com/user-attachments/assets/d7ba70cf-c9af-4901-91a7-3cdbe0851077)

Log: 
![Screenshot 2024-10-09 at 12 48 09 PM](https://github.com/user-attachments/assets/73355113-75a1-484e-8117-44834fb4f6a4)

laceysanderson committed:
Update formats to match field names.

laceysanderson committed:
Apply suggestions from code review

Co-authored-by: Carolyn T Caron <carolyn.caron@usask.ca>

laceysanderson committed:
Switch germplasm species to germplasm scientific name.

laceysanderson committed:
Added method as discussed by myself and Ruobin.

laceysanderson commented:
Thanks for all the suggestions @carolyncaron! 

> I think it would be helpful to bold text that reminds the user to add another item when there are multiple

Unfortunately Drupal will only let me use HTML in field group descriptions, not in field descriptions so I am unable to make this change 😢 I agree wholeheartedly!

> Regarding Germplasm genus and species - what happens when we have more than one genus (Lens + Phaseolus) and then multiple species for each?

I updated the species field to Scientific Name as was suggested during in an earlier offline conversion. This way there is no ambiguity.


**Note: I also added a method field for the phenotypes as this came up in coversation with Ruobin at one point as being needed.**

laceysanderson committed:
Merge remote-tracking branch 'origin/4.x' into g1.25-experimentContentTypeUpdates

laceysanderson committed:
Small updates.

laceysanderson committed:
Snuck in during merge conflict resolution.

laceysanderson committed:
Update Drupal version

laceysanderson committed:
Remove apply update since it appears not to be needed.

laceysanderson commented:
Note: With the merge of PR #28, we can now create experiments again!

**However, you need to make sure it's not using an old version of Tripaldocker.** To do this I suggest just deleting all the containers/images depending on Tripaldocker, then delete the tripaldocker image. Now when you rebuild it will pull the most recent version of Tripaldocker and you will be able to create an experiment!

ruobinLiu commented:
> > Regarding Germplasm genus and species - what happens when we have more than one genus (Lens + Phaseolus) and then multiple species for each?
> 
> I updated the species field to Scientific Name as was suggested during in an earlier offline conversion. This way there is no ambiguity.
![Screenshot 2024-10-11 at 4 10 43 PM](https://github.com/user-attachments/assets/fd19c95b-e94e-4a4e-8534-26593ed98539)

I've done test on this part, the result turns out pretty clear to me !

carolyncaron reviewed (changes requested):
I confirmed that I am able to create experiment pages now! 🎉 

I went back to edit my page a few times and usually the changes showed up as expected. The exception is when I started adding fields for additional experiment types. 😢 It seems that when a field is similar enough to another one in a different Experiment type, one of them will override the other one. This occurred in my testing between:
- **Field Locations**, **Greenhouse Locations** and **Growth Chamber Locations**
- **Field Experiment Design** and both **Experimental Design** fields under Greenhouse and Growth Chamber
- **Growing Media** under Greenhouse and Growth Chamber

You can see that these fields have the same values in the following screenshot, despite that I originally wrote different values (it's also pretty random whether the newly added one overrides the field that was already filled out, or vice versa):
![localhost_experiment (3)](https://github.com/user-attachments/assets/bd023b33-24d4-43b2-bdda-8d746fd3f8bc)

When I go to edit again, the editable values appear as they do on the page. 

I wanted to see what happens when I add additional items for each of the Location fields and the issue is consistent (in this case, I noticed that the Greenhouse Location of **Rosthern, SK, Canada** is the one that overwrote the other two additional fields):
<img width="689" alt="Screenshot 2024-10-11 at 4 22 24 PM" src="https://github.com/user-attachments/assets/650d2821-da50-41ff-aa50-bb76b62340e9">

carolyncaron commented:
For **Experiment Date** under Biochemical Assay, I put in the example given:
<img width="481" alt="Screenshot 2024-10-11 at 4 29 08 PM" src="https://github.com/user-attachments/assets/1495bff7-d718-4b66-8c3e-c3f75dfba54a">

And it displays as "2021":
<img width="297" alt="Screenshot 2024-10-11 at 4 31 04 PM" src="https://github.com/user-attachments/assets/60ec31a3-a960-48f8-958f-67e1fc5a1d7d">

ruobinLiu commented:
I also confirmed that I am able to create experiment pages!
The Overview, Germplasm, Phenomic sections all very intuitive and straight. I like how it allows using individual years without having to stuck with a range of years.
![Screenshot 2
![Screenshot 2024-10-11 at 5 13 57 PM](https://github.com/user-attachments/assets/1671d313-c923-4c0b-a6bd-873ebf145b98)
024-10-11 at 5 13 46 PM](https://github.com/user-attachments/assets/118512be-5e46-4ac3-b28a-0147992b3e49)

ruobinLiu commented:
As Carolyn mentioned, I also experienced issues with Experimental Design section. More specifically, 

1. Field Locations, Greenhouse Locations and Growth Chamber Locations force to present the same values even if I have given them different ones; plus, I wasn't able to go back to the edit and manually fix them.
![Screenshot 2024-10-11 at 5 19 32 PM](https://github.com/user-attachments/assets/fca3414f-0330-47f0-9de3-be76e08bbb3d)
2. Biochemical Assay therefore, has no problem to show the values I put, but like Carolyn has already mentioned, Experiment Date part only has the year display, without fully showing the date and time I've put.
![Screenshot 2024-10-11 at 5 24 10 PM](https://github.com/user-attachments/assets/ce6ec615-638a-4f3b-a608-a5e7fcae7480)
![Screenshot 2024-10-11 at 5 27 50 PM](https://github.com/user-attachments/assets/09506264-c967-4679-aecf-7659fc4036d3)

laceysanderson committed:
Merge remote-tracking branch 'origin/4.x' into g1.25-experimentContentTypeUpdates

laceysanderson committed:
Fix modules required now that Content Types is a kernel test.

laceysanderson committed:
Only check/fix on save please.

laceysanderson committed:
Add in 3rd party settings needed for publish

laceysanderson committed:
Switch the form to vertical tabs and support the advanced sidebar.

laceysanderson committed:
Install markup module and it's patch.

laceysanderson committed:
Merge branch 'g1.25-experimentContentTypeUpdates' of https://github.com/TripalCultivate/TripalCultivate into g1.25-experimentContentTypeUpdates

laceysanderson committed:
Start working on layout changes.

laceysanderson committed:
Add some form explanation text files -not used yet.

laceysanderson committed:
Organize the objectives/summary step.

laceysanderson committed:
Add exported fields --non-tripal only

laceysanderson committed:
Create Markup fields programatically

laceysanderson committed:
Remove weights so that the order in the layout is respected.

laceysanderson committed:
Ensure the markup field include the instructions from the twig files.

laceysanderson committed:
Add support for creating entity reference fields.

laceysanderson committed:
Clean-up / organize the attribution step.

laceysanderson committed:
Merge branch '4.x' into g1.25-experimentContentTypeUpdates

laceysanderson committed:
Start with the germplasm + experiment design sections.

laceysanderson committed:
Start organizing design metadata (e.g. locations) to be less experiment type focused.

laceysanderson committed:
Merge branch '4.x' into g1.25-experimentContentTypeUpdates

laceysanderson committed:
Firm up the Experiment Design Section.
